sensing/autoware_cuda_pointcloud_preprocessor/include/autoware/cuda_downsample_filter/cuda_voxel_grid_downsample_filter.hpp

```diff
@@ -0,0 +1,137 @@
+// Copyright 2025 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef AUTOWARE__CUDA_DOWNSAMPLE_FILTER__CUDA_VOXEL_GRID_DOWNSAMPLE_FILTER_HPP_
+#define AUTOWARE__CUDA_DOWNSAMPLE_FILTER__CUDA_VOXEL_GRID_DOWNSAMPLE_FILTER_HPP_
+
+#include "autoware/cuda_downsample_filter/thrust_custom_allocator.hpp"
+
+#include <cuda_blackboard/cuda_pointcloud2.hpp>
+#include <cuda_blackboard/cuda_unique_ptr.hpp>
+
+#include <cuda_runtime.h>
+
+#include <cstdint>
+#include <iostream>
+#include <memory>
+
+#define CUDA_VGDF_ERROR_CHECK(e) (CUDA_BLACKBOARD_CHECK_CUDA_ERROR(e))
+
+namespace autoware::cuda_downsample_filter
+{
+
+class CudaVoxelGridDownsampleFilter
+{
+public:
+  template <typename T>
+  struct ThreeDim
+  {
+    T x;
+    T y;
+    T z;
+
+    friend std::ostream & operator<<(std::ostream & os, const ThreeDim & t)
+    {
+      os << t.x << ", " << t.y << ", " << t.z;
+      return os;
+    }
+
+    T & operator[](size_t i)
+    {  // Define [] operator to make iterative access easy
+      static T * members[] = {&x, &y, &z};
+      return *members[i];
+    }
+  };
+
+  struct OptionalField
+  {
+    bool is_valid;
+    size_t offset;
+  };
+
+  struct VoxelInfo
+  {
+    size_t num_input_points;
+    size_t input_point_step;
+    size_t input_xyzi_offset[4];
+    OptionalField input_return_type_offset;
+    OptionalField input_channel_offset;
+    size_t output_offsets[6];
+    ThreeDim<float> voxel_size;
+    ThreeDim<int> min_coord;
+    ThreeDim<int> max_coord;
+  };
+
+  // This filter only returns XYZIRC regardless of the input point type
+#pragma pack(push, 1)
+  struct OutputPointType
+  {
```

Comment:
You should be able to use https://github.com/autowarefoundation/autoware_universe/blob/d82b071a2aca41c9a2be5be964d16b0594351ee5/sensing/autoware_cuda_pointcloud_preprocessor/include/autoware/cuda_pointcloud_preprocessor/point_types.hpp#L23
If you are worried about packing (the structure was aligned though) you can add the macro there

Reply:
Thanks for pointing this out. I modified the code to use the struct you suggested in 76a4f76ee